contrib/platform/src/com/sun/jna/platform/win32/Kernel32Util.java

```diff
@@ -678,4 +680,141 @@ public static final String extractVolumeGUID(String volumeGUIDPath) {
+	 *            The name or ID of the resource
+	 * @return The resource bytes, or null if no such resource exists.
+	 */
+	public static byte[] getResource(String path, String type, String name) {
+		final HMODULE target = Kernel32.INSTANCE.LoadLibraryEx(path, null, Kernel32.LOAD_LIBRARY_AS_DATAFILE);
+
+		Pointer t = null;
+		try {
+			t = new Pointer(Long.parseLong(type));
+		} catch (Exception e) {
+			t = new Memory(Native.WCHAR_SIZE * (type.length() + 1));
+			t.setWideString(0, type);
+		}
+
+		Pointer n = null;
+		try {
+			n = new Pointer(Long.parseLong(name));
+		} catch (Exception e) {
+			n = new Memory(Native.WCHAR_SIZE * (name.length() + 1));
+			n.setWideString(0, name);
+		}
+
+		HRSRC hRsrc = Kernel32.INSTANCE.FindResource(target, n, t);
+		if (hRsrc == null) {
+			return null;
+		}
+
+		HANDLE loaded = Kernel32.INSTANCE.LoadResource(target, hRsrc);
+		int length = Kernel32.INSTANCE.SizeofResource(target, hRsrc);
+		Pointer start = Kernel32.INSTANCE.LockResource(loaded);
+		return start.getByteArray(0, length);
+	}
+
+	/**
+	 * Gets a list of all resources from the specified executable file
+	 * 
+	 * @param path
+	 *            The path to the executable file
+	 * @return A map of resource type name/ID => resources.<br>
+	 *         A map key + a single list item + the path to the executable can
+	 *         be handed off to getResource() to actually get the resource.
+	 */
+	public static Map<String, List<String>> getResourceNames(String path) {
+		final HMODULE target = Kernel32.INSTANCE.LoadLibraryEx(path, null, Kernel32.LOAD_LIBRARY_AS_DATAFILE);
+
+		final List<String> types = new ArrayList<String>();
+		WinBase.EnumResTypeProc ertp = new WinBase.EnumResTypeProc() {
+
+			@Override
+			public boolean invoke(HMODULE module, Pointer type, Pointer lParam) {
+				// simulate IS_INTRESOURCE
+				if (Pointer.nativeValue(type) < 65535) {
+					types.add(Pointer.nativeValue(type) + "");
+				} else {
+					types.add(type.getWideString(0));
+				}
+				return true;
+			}
+		};
+
+		Kernel32.INSTANCE.EnumResourceTypes(target, ertp, null);
+
+		final Map<String, List<String>> result = new LinkedHashMap<String, List<String>>();
+
+		for (final String typeName : types) {
+			result.put(typeName, new ArrayList<String>());
+
+			Pointer pointer = null;
+			try {
+				pointer = new Pointer(Long.parseLong(typeName));
+			} catch (Exception e) {
+				pointer = new Memory(Native.WCHAR_SIZE * (typeName.length() + 1));
+				pointer.setWideString(0, typeName);
+			}
+
+			Kernel32.INSTANCE.EnumResourceNames(target, pointer, new WinBase.EnumResNameProc() {
+
+				@Override
+				public boolean invoke(HMODULE module, Pointer t, Pointer name, Pointer lParam) {
+					// simulate IS_INTRESOURCE
+					if (Pointer.nativeValue(name) < 65535) {
+						result.get(typeName).add(Pointer.nativeValue(name) + "");
+					} else {
+						result.get(typeName).add(name.getWideString(0));
+					}
+
+					return true;
+				}
+			}, null);
+		}
+
+		return result;
```

Comment:
The whole method should be braced within a _try-finally_ block:
```java
HMODULE module = ...LoadLibraryEx...
if (null or INVALID) {
    throw Win32Exception(...could not load ...: GetLastError());
}
try {
} finally { // avoid handle leakage
    /*
     * TODO consider implementing a mechanism that detects whether we are
     * exiting due to an exception and make the original exception suppress this
     * one instead of the other way around (which is what the simple code does).
     * If you can't figure out a simple way how to do this let me know and I will show you
     */
    if (!FreeLibrary(module)) {
        throw new Win32Exception(...failed to free: GetLastError()...);
    }
}
```

To quote from https://msdn.microsoft.com/en-us/library/windows/desktop/ms684179(v=vs.85).aspx:
>Use the FreeLibrary function to free a loaded module, whether or not loading the module caused its reference count to be incremented. If the module was loaded as a data or image file, the mapping is destroyed but the reference count is not decremented. Otherwise, the DLL reference count is decremented. Therefore, it is safe to call FreeLibrary with any handle returned by LoadLibraryEx.